aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp

```diff
@@ -299,28 +323,45 @@ Tensor _qembeddingbag_nbit_prepack_helper(
 
 #ifdef USE_FBGEMM
   if (!optimized_qparams) {
-    at::parallel_for(
-      0, embedding_rows, 1, [&](int32_t start_idx, int32_t end_idx) {
-        for (int64_t row = start_idx; row < end_idx; ++row) {
-          fbgemm::FloatOrHalfToFusedNBitRowwiseQuantizedSBHalf<float>(
-            bit_width, weight_data + row * embedding_cols, 1,
-            embedding_cols, output_data + row * output_shape[1]);
-        }
-      });
+    if (weight.scalar_type() == at::ScalarType::Half) {
+      const auto weight_data = static_cast<fbgemm::float16*>(weight.data_ptr());
+      at::parallel_for(
+        0, embedding_rows, 1, [&](int32_t start_idx, int32_t end_idx) {
+          for (int64_t row = start_idx; row < end_idx; ++row) {
+            fbgemm::FloatOrHalfToFusedNBitRowwiseQuantizedSBHalf<fbgemm::float16>(
+              bit_width, weight_data + row * embedding_cols, 1,
+              embedding_cols, output_data + row * output_shape[1]);
+          }
+        });
+    }
+    else {
+      const auto weight_data = weight.data_ptr<float>();
+      at::parallel_for(
+        0, embedding_rows, 1, [&](int32_t start_idx, int32_t end_idx) {
+          for (int64_t row = start_idx; row < end_idx; ++row) {
+            fbgemm::FloatOrHalfToFusedNBitRowwiseQuantizedSBHalf<float>(
+              bit_width, weight_data + row * embedding_cols, 1,
+              embedding_cols, output_data + row * output_shape[1]);
+          }
+        });
+    }
   } else {
 #endif // USE_FBGEMM
     const auto output_columns = output.size(output.dim() - 1);
 
     for (int row = 0; row < embedding_rows; ++row) {
-      const float* input_row = weight_data + row * embedding_cols;
+      const auto row_tensor = weight_contig.scalar_type() == at::ScalarType::Half
+        ? weight_contig[row].to(at::ScalarType::Float)
+        : weight_contig[row];
```

Comment:
Can we move the casting outside the for-loop? similar to `qembeddingbag_byte_prepack` above?